V1.10 (#428)
mulesoft/docs-runtime-fabric: modules/ROOT/pages/custom-ingress-configuration.adoc

= Configure Ingress for Runtime Fabric on Self-Managed Kubernetes
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

:page-aliases: enable-inbound-traffic-self.adoc

[IMPORTANT]
====

> This content refers to configuring ingress on Runtime Fabric for versions 1.10 and later. Refer to the documentation for versions 1.8 or 1.9 for previous ingress configuration instructions.

Reviewer: Can we hyperlink the keys `1.8` and `1.9` ?

Author: I'll leave this for Isaac to decide when he gets back.

==== 

> Anypoint Runtime Fabric on Self-Managed Kubernetes enables you to specify custom Ingress configurations using an ingress resource template. In a template, similar to a https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes ingress spec^], you can specify annotations, `ingressClassName`, and HTTP and HTTPS rules. 

Reviewer: Why is just `ingressClassName` highlighted?

Reviewer: ```suggestion
Anypoint Runtime Fabric on Self-Managed Kubernetes enables you to specify custom Ingress configurations using an ingress resource template. In a template, similar to a https://kubernetes.io/docs/concepts/services-networking/ingress/[Kubernetes ingress spec^], you can specify annotations, ingress class name, and HTTP and HTTPS rules. 
```

. The IT admin applies the template using `kubectl apply` commands.
. From the applied template, Runtime Fabric creates placeholder URI domains and propagates the ingress configuration in Anypoint Runtime Manager as a template for all application deployments.
. A Mule application developer uses Runtime Manager to configure an application for deployment by selecting an available host, an optional subdomain, and a path combination, as specified in the ingress template. 
. Runtime Fabric receives the application deployment request and uses the ingress configurations from the template to create a corresponding ingress object in the cluster.

== Example Ingress Resource Templates

The following examples illustrate how to modify a Kubernetes ingress spec to create a Runtime Fabric ingress resource template. 

Use the API version appropriate to your Kubernetes version:

* Kubernetes v1.19 or later: 
** networking.k8s.io/v1 
** networking.k8s.io/v1beta1
+
As of v1.22, the `extensions/v1beta1` and `networking.k8s.io/v1beta1` API versions of Ingress are no longer available.

> * Kubernetes v1.14 through v1.18: 

Reviewer: We officially dont support anything below K8 v1.17

Reviewer: ```suggestion
* Kubernetes version v1.18 and earlier: 
```

** only networking.k8s.io/v1beta1  

=== networking.k8s.io/v1

[NOTE]
====
This template is a simplified version of the configuration. The final configuration of the template can vary depending on the controller and ingress routing rules you use. Review the documentation for your ingress controller carefully.
====

[source,yaml]
--
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: example-rtf-ingress
  namespace: rtf #<1>                           
  labels:
    business-group: business-group-id
    environment: environment-id
spec:
  ingressClassName: rtf-nginx #<2> 
  tls: #<4>
  - hosts:
      - app-name.example.com
    secretName: example-tls
  rules:
  - host: app-name.example.com #<3>
    http:
      paths: #<5>
      - pathType: Prefix
        path: /
        backend: #<6>
          service:
            name: service-name
            port:
              name: service-port
--

Note the following about this example:
  
<1>  The template must be placed in the `rtf` namespace.

<2>  ingressClassName must be prefixed with `rtf-`, for example, `rtf-nginx`.
+

> This is how Runtime Fabric recognizes the object as a template. A template that uses the `rtf-` prefix in the `ingressClassName`, for example, `rtf-nginx`, is consumed by the Runtime Fabric agent only and not by the actual ingress controller. Ingress controllers discover only resources with an `ingressClassName` value that uses the vendor-specific name, for example, `nginx` or `haproxy`.

Reviewer: Should we include the `ingress.class` in annotations ? @kmendiratta-crm

Reviewer: ```suggestion
This is how Runtime Fabric recognizes the object as a template. A template that uses the `rtf-` prefix in the `spec.ingressClassName` field or in the `kubernetes.io/ingress.class` annotation, for example, `rtf-nginx`, is consumed by the Runtime Fabric agent only and not by the actual ingress controller. Ingress controllers discover only resources with an `spec.ingressClassName` value that uses the vendor-specific name, for example, `nginx` or `haproxy`.
```

> <3> TLS is optional.

Reviewer: It will be worth mentioning that:
If the TLS section is added with a "secretName" its the responsibility of the user to make sure that the referred tls secret is created in the RTF namespace.

Reviewer: ```suggestion
<3> TLS is optional.
+
If the TLS section is added with a `secretName` its the responsibility of the user to make sure that the referred TLS secret is created in the RTF namespace.
```

<4> `app-name` is a placeholder parameter that will be replaced with actual app name when you deploy the application. It helps to ensure that each endpoint name is unique if there is not a wildcard in the subdomain.

<5> A template can include multiple hosts, but Runtime Manager displays the first path rule only.

<6>  These are placeholder values required for Kubernetes validation, but the actual values are not used by Runtime Fabric. 

For a path-based routing ingress template, consult your ingress controller documentation and configure it accordingly.
For an example using Nginx, see https://help.mulesoft.com/s/article/Path-based-Nginx-Ingress-Routing-Example-for-Runtime-Fabric-on-Self-Managed-Kubernetes[Path-based Nginx Ingress Routing Example for Runtime Fabric on Self-Managed Kubernetes].

=== networking.k8s.io/v1beta1

[source,yaml]
--

> apiVersion: networking.k8s.io/v1beta1

Reviewer 2: ```suggestion
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: example-rtf-ingress
  namespace: rtf
  labels:
    business-group: business-group-id
    environment: environment-id
spec:
  ingressClassName: rtf-nginx
  tls:
  - hosts:
      - app-name.example.com
    secretName: example-tls
  rules:
    - host: app-name.example.com
      http:
        paths:
        - path: /
          pathType:
          backend:
            serviceName: service-name
            servicePort: 80
```

kind: Ingress
  - host: app-name.example.com
    http:
      paths: #<1>
      - path: /
        pathType: 
        backend: #<2>
          serviceName: service

>           servicePort: 80

Reviewer 2: This example is copy-paste accurate, i.e. if a customer copy pastes and attempts to kubectl apply this, it will fail.

. Select *Ingress*.
. From the *Host* drop-down list, select a host for the application. 
. If the hostname uses a wildcard, add a subdomain in the *Subdomain* field. 
+
The *Subdomain* field is only available if the hostname uses a wildcard.
. In the *Path* field, add a URL path to the application’s endpoint:
+
image::rtf-ingress-endpoint.png[The host and path fields are populated for the endpoint,80%]
+
It can be any of the placeholder parameters, such as `app-name`.
. To preview the endpoint, click the generated preview link.
. To add additional endpoints, click *+ Add Endpoint*.
. When ready, click *Deploy application*.  

== Existing Ingress Configurations Translated Post-Upgrade 

When you upgrade to Runtime Fabric version 1.10, the upgrade process automatically translates any existing ingress ConfigMaps to the ingress resource template format, including annotations, path, namespace, and host parameters. Review the table below to determine if any action is required after you upgrade.

[%header%autowidth.spread]
|===
|Runtime Fabric Pre-upgrade |Runtime Fabric Upgraded |Action Required |TLS Support

> |No domains configured and no ingress-ConfigMap applied |Generated template will contain `*` for hosts and will include `/app-name` in the path |Edit the `host` field in the generated template to provide a valid host/domain for deploying applications in Runtime Manager|No TLS section added by default 

Reviewer 2: ```suggestion
|No domains configured and no ingress-ConfigMap applied |No template will be generated automatically after upgrade|Create a template as described above |-
```

+
`kubectl get ingress -n [NAMESPACE]`

. Validate that the service was created:
+
`kubectl get svc -n [NAMESPACE]`
+
If not, check the Runtime Fabric agent logs:
+
`kubectl logs -n rtf [AGENT_POD_NAME] -f`

If the service and ingress object look healthy, see additional troubleshooting tasks. 

=== Ingress Resource Created in Cluster But Cannot Access Its Endpoint

Scenario: Runtime Fabric successfully creates an ingress resource in the cluster but you cannot access the application’s endpoint due to a 404 error. 

To troubleshoot this issue: 

. Verify the ingress and service resources:
+
----
kubectl get ingress -n<APP-NAMESPACE>
kubectl get svc -n<APP-NAMESPACE>
----
+
The results should be similar to:
+
----
# kubectl get ingress -n<<APP-NAMESPACE>                             
NAME                           CLASS       HOSTS               ADDRESS       PORTS     AGE
<INGRESS_RESOURCE>            <INGRESS>    <HOSTNAME>.com      <HOST IP>     80, 443   7m3s

#kubectl get svc -n<<APP-NAMESPACE>                                
NAME             TYPE        CLUSTER-IP     EXTERNAL-IP   PORT(S)                      AGE

> test-rc-1        ClusterIP   <CLUSTER IP>   <none>        8081/TCP,8082/TCP,5701/TCP   8m5s

Reviewer 2: ```suggestion
<SERVICE-NAME>       ClusterIP   <CLUSTER IP>   <none>        8081/TCP,8082/TCP,5701/TCP   8m5s
```

    http:
      paths: #<5>
      - pathType: Prefix
        path: /
        backend: #<6>
          service:
            name: service-name
            port:
              name: service-port
--

Note the following about this example:
  
<1> The template must be placed in the `rtf` namespace.

<2> ingressClassName must be prefixed with `rtf-`, for example, `rtf-nginx`.
+
Runtime Fabric uses the `rtf-` prefix to recognize the object as a template.
Templates with `rtf-` prefix in the `spec.ingressClassName` field or in the `kubernetes.io/ingress.class` annotation
are consumed by the Runtime Fabric agent only and not by the actual ingress controller.
Ingress controllers discover only those resources with a `spec.ingressClassName` value that use the vendor-specific name, for example, `nginx` or `haproxy`.

<3> TLS is optional.
+
If you add the TLS section with a `secretName` attribute, you must ensure that you also create the referred TLS secret in the `rtf` namespace.

<4> The `app-name` placeholder parameter will be replaced with actual app name when you deploy the application. It helps to ensure that each endpoint name is unique if there is not a wildcard in the subdomain.

> <5> A template can include multiple hosts, but Runtime Manager displays only the first path rule.

Reviewer: ```suggestion
<5> A template can include multiple paths for a host, but Runtime Manager displays only the first path rule for the host.
```

For a path-based routing ingress template, consult your ingress controller documentation and configure it accordingly.
For an example using Nginx, see https://help.mulesoft.com/s/article/Path-based-Nginx-Ingress-Routing-Example-for-Runtime-Fabric-on-Self-Managed-Kubernetes[Path-based Nginx Ingress Routing Example for Runtime Fabric on Self-Managed Kubernetes].

=== networking.k8s.io/v1beta1

[source,yaml]
--
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: example-rtf-ingress
  namespace: rtf
  labels:
    business-group: business-group-id
    environment: environment-id
spec:
  ingressClassName: rtf-nginx
  tls:
  - hosts:
      - app-name.example.com
    secretName: example-tls
  rules:
    - host: app-name.example.com
      http:
        paths:
        - path: /
          pathType:
          backend:
            serviceName: service-name
            servicePort: 80
kind: Ingress
  - host: app-name.example.com
    http:
      paths: #<1>
      - path: /
        pathType: 
        backend: #<2>
          serviceName: service
          servicePort: 80
--

Note the following about this example:

> <1> A template can include multiple hosts, but Runtime Manager displays the only first path rule.

Reviewer: ```suggestion
<1> A template can include multiple paths for a host, but Runtime Manager displays only the first path rule for the host.
```
